builder/googlecompute/step_import_os_login_ssh_key.go

```diff
@@ -0,0 +1,92 @@
+package googlecompute
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+// StepCreateSSHKey represents a Packer build step that generates SSH key pairs.
+type StepImportOSLoginSSHKey struct {
+	Debug bool
+}
+
+// Run executes the Packer build step that generates SSH key pairs.
+// The key pairs are added to the ssh config
+func (s *StepImportOSLoginSSHKey) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(*Config)
+	driver := state.Get("driver").(Driver)
+	ui := state.Get("ui").(packer.Ui)
+
+	if !config.UseOSLogin {
+		return multistep.ActionContinue
+	}
+
+	ui.Say("Importing SSH public key for OSLogin...")
+
+	// Generate SHA256 fingerprint of SSH public key
+	// Put it into state to clean up later
+	sha256sum := sha256.Sum256(config.Comm.SSHPublicKey)
+	state.Put("ssh_key_public_sha256", hex.EncodeToString(sha256sum[:]))
+
+	// TODO: @cpwc need to obtain user's email
+	if config.account == nil {
+		err := fmt.Errorf("Error importing SSH public key for OSLogin: need user's ID/email")
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	loginProfile, err := driver.ImportOSLoginSSHKey(config.account.Email, string(config.Comm.SSHPublicKey))
+	if err != nil {
+		err := fmt.Errorf("Error importing SSH public key for OSLogin: %s", err)
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	// Setting/replacing SSHUsername as the username have to be from OSLogin
+	if len(loginProfile.PosixAccounts) == 0 {
+		err := fmt.Errorf("Error importing SSH public key for OSLogin: no PosixAccounts available")
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+	ui.Say("Setting SSH Username for OSLogin...")
+	if s.Debug {
+		ui.Message(fmt.Sprintf("ssh_username: %s", loginProfile.PosixAccounts[0].Username))
+	}
+	config.Comm.SSHUsername = loginProfile.PosixAccounts[0].Username
```

Comment:
@nywilken yep I agree with you on the approach to get the best possible possible account to be used as any `PosixAccount` once given the appropriate `oslogin.role` should be able to SSH into the system. Additional checks should be included if we have to select `windows` or `linux`. But currently only `linux`.

>One open question I have is: When a new SSH public key is added can it be used with any of the configured Posix Accounts?

Yes. That's right. 

>Add two new configuration options that are required if use_os_login is true. This would also solve the account file requirement. Which is only an issue for users using service accounts, correct?

This will solve the account file requirement actually not specific to service account.

Hopefully this make sense.

Reply:
Hi @cpwc apologies in advance for all the long comments; I'm still trying to understand what all the docs are saying so things are fuzzy. I've essentially considered this a GitHub pairing session to make sense of the details. Which means we should test our assumptions and any new findings :smile: 

I spent a little more time looking into the PosixAccounts because I was curious to know when and how a user can have multiple accounts. I found this interesting issue which seems to imply that a PosixAccount created by Os Login will always be the same as it is tied to a user's Google identity - https://github.com/GoogleCloudPlatform/compute-image-packages/issues/692

Looking at the [documentation](https://cloud.google.com/compute/docs/oslogin/#benefits_of_os_login) a little more carefully it also confirms what is said in the issue that only one account will be tied to the user. 

Now in the case where a user's login information is being synced from AD or LDAP I think the Linux information may be different. In that case I would think the user would just use the `ssh_username` and `ssh_private_key_file` to  connect to the instance with no `use_os_login` set because it may already be set on the project level. We need a test account for this. 

So what I am really trying to say is that after further review I think that your initial implementation of choosing `loginProfile.PosixAccounts[0].Username` should be enough. 


> Add two new configuration options that are required if use_os_login is true. This would also solve the account file requirement. Which is only an issue for users using service accounts, correct?

Given the info above I think this may not work like we think as the username will always be the same and the email must be the one tied to the Google account identity (domain). Please hold on adding any additional options. 

> This will solve the account file requirement actually not specific to service account.

Sorry my question wasn't clear. What I meant to ask is "what are the use cases where `use_os_login` can be enabled and an account file is not needed"? 

As I understand it, the Packer builder requires an account file for authentication unless it is running on a GCE instance. When running on a GCE instance it relies on the service account associated with the instance for authentication. But I might be missing other uses cases.

Reply:
Hey @nywilken no worries. I'll love to iron this out together. 

>Looking at the documentation a little more carefully it also confirms what is said in the issue that only one account will be tied to the user.

That being said tho. What if the use have multiple alias attached to it? Will he/she have multiple PosixAccounts? I've not verified this tho as I do not have access to create alias for the domain I'm currently using. 

>So what I am really trying to say is that after further review I think that your initial implementation of choosing loginProfile.PosixAccounts[0].Username should be enough.

Based on your findings and once we are able to verify above that it is not true. I think it's safe enough for us to conclude that the initial implementation is good enough.

>Given the info above I think this may not work like we think as the username will always be the same and the email must be the one tied to the Google account identity (domain). Please hold on adding any additional options.

Sure! Will hold my changes on that for now.

>As I understand it, the Packer builder requires an account file for authentication unless it is running on a GCE instance. When running on a GCE instance it relies on the service account associated with the instance for authentication. But I might be missing other uses cases.

Ah. If I'm running locally on my machine itself with gcloud SDK configured properly I do not need the `account_file`. As the SDK itself will fallback to Application Default Credentials. Do correct me if I'm wrong as that's my understanding [here](https://github.com/googleapis/google-cloud-go#authorization).

Reply:
> That being said tho. What if the use have multiple alias attached to it? Will he/she have multiple PosixAccounts? I've not verified this tho as I do not have access to create alias for the domain I'm currently using.

So I reached out to some Googlers to see what I can find and the guidance is that an alias will most likely roll up into the primary account. The suggested path forward is to filter for the Primary account and return that one as the selected username - improvement to the suggestion I posted above. 


The idea being if  only one account is returned it will be the Primary username and if there are multiple accounts selecting the Primary should give a reasonable behavior. I agree with the path forward and if we find we need to change things in the future we will handle it then.

Reply:
> Ah. If I'm running locally on my machine itself with gcloud SDK configured properly I do not need the account_file. As the SDK itself will fallback to Application Default Credentials. Do correct me if I'm wrong as that's my understanding here.

Ah okay. This is interesting because the Application Default Credentials is just an env variable that points to the same account file that you would specify in  the `account_file` builder configuration option. I can build the same images by just setting the env var, but on this branch it fails because account is never populated. I wonder the builder needs to set the account details if the date is read from the env var. Let me dive into this.

Reply:
>So I reached out to some Googlers to see what I can find and the guidance is that an alias will most likely roll up into the primary account. The suggested path forward is to filter for the Primary account and return that one as the selected username - improvement to the suggestion I posted above.

>The idea being if only one account is returned it will be the Primary username and if there are multiple accounts selecting the Primary should give a reasonable behavior. I agree with the path forward and if we find we need to change things in the future we will handle it then.

Awesome! Looks like we can agree on picking up the `primary` account for now and improve it in the future if required. Let me pick it up and update it.

>I can build the same images by just setting the env var, but on this branch it fails because account is never populated. I wonder if the builder just needs to set the account details if the account data is read from the env var. Let me dive into this.

Thanks for helping to deep dive on this one @nywilken appreciate it 🙏